refactor volumes page into core, external

geoffcline opened the pull request:
This PR is a rewrite of the concepts/storage/volumes page. I've reorganized it for improved readability.


re: Create two umbrella sections (persistent, ephemeral) for volume concept page #24726 

The previous list of volume types included a lot of different things. 

It's now broken up into two groups. I went with "core" and "external" -- but I'm open to feedback.

geoffcline commented:
remaining tasks

- [ ] ms spell check
- [ ] double check no conflicting header links
- [ ] double check style guide compliance re: capitalization

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
 
-To disable the `awsElasticBlockStore` storage plugin from being loaded by the controller manager
-and the kubelet, set the `CSIMigrationAWSComplete` flag to `true`. This feature requires the `ebs.csi.aws.com` Container Storage Interface (CSI) driver installed on all worker nodes.
+Kubernetes implements the Container Storage Interface (CSI). CSI is the modern way to communicate with out-of-tree plugins for externaly managed volumes. More specifically, CSI is a standard for exposing arbitrary block and file storage systems to containerized workloads. Learn more with the [CSI docs](https://kubernetes-csi.github.io/docs/).
```
I'd write:
```suggestion
Kubernetes implements the {{< glossary_tooltip text="Container Storage Interface" term_id="csi" >}} (CSI),
which lets developers implement storage plugins without needing to make changes to the code of
Kubernetes itself. Visit
[Kubernetes Container Storage Interface Documentation])(https://kubernetes-csi.github.io/docs/)
to learn more about CSI.
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
+*Out-of-tree* storage plugins are not part of the kubernetes binary. 
 
-The `azureDisk` volume type mounts a Microsoft Azure [Data Disk](https://docs.microsoft.com/en-us/azure/aks/csi-storage-drivers) into a pod.
+`FlexVolume` is an alternate way to communicate with out of tree volume plugins. `FlexVolume` was developed before CSI.
```
```suggestion
[FlexVolume](#flexvolume) is an another mechanism for storage plugins. Unlike CSI, FlexVolume is
more closely tied to Kubernetes and its storage implementation. FlexVolume was developed before CSI.
```
I would not write `FlexVolume`, though, because when readers are specifying it as a volume type they are most likely to write `flexVolume`. To me, using FlexVolume just as a proper noun is OK.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
+`FlexVolume` is an alternate way to communicate with out of tree volume plugins. `FlexVolume` was developed before CSI.
 
-For more details, see the [`azureDisk` volume plugin](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/staging/volumes/azure_disk/README.md).
+   * [CSI](#CSI)
+   * [FlexVolume](#flexVolume)
+   * [persistentVolumeClaim](#persistentvolumeclaim)
```
These aren't really rival technologies; PVCs are relevant to both CSI and FlexVolume. I would omit this list.

geoffcline replied:
I think it's important to keep this list. It's at least a listing of the storage plugin related page sections. I refreshed the description here to indicate that PVCs are a generic way to consume different storage plugins.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
-#### azureDisk CSI migration
 
-{{< feature-state for_k8s_version="v1.19" state="beta" >}}
+### External Volume Types
```
I'm not sure what this heading covers. I'd expect to see `cephfs`, `gitRepo`, `glusterfs`, `iscsi`, `nfs` and `rbd` listed in the same group as `gcePersistentDisk`, `cinder` and the others here.

geoffcline replied:
The goal here was to include externally _managed_ volumes, such as storage as a service from cloud providers. It's a hazier distinction that I expected. I commented above about the category location for specific types.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
+other containers in the same pod, or even to other pods on the same node.
 
-{{< feature-state for_k8s_version="v1.15" state="alpha" >}}
+Mount propagation of a volume is controlled by the `mountPropagation` field in `Container.volumeMounts`.
```
```suggestion
Mount propagation of a volume is controlled by the `mountPropagation` field within `volumeMounts`.
```
?

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
-Driver](https://github.com/kubernetes-sigs/azurefile-csi-driver)
-must be installed on the cluster and the `CSIMigration` and `CSIMigrationAzureFile`
-alpha features must be enabled.
+* `None` - This volume mount will not receive any subsequent mounts
```
```suggestion
* `None` - This volume mount does not receive any subsequent mounts
```
Also I'd mention that it's default just after `None`.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
-unmounted. This means that a `cephfs` volume can be pre-populated with data, and
-that data can be shared between pods. The `cephfs` volume can be mounted by multiple
-writers simultaneously.
+* `HostToContainer` - This volume mount will receive all subsequent mounts
```
```suggestion
* `HostToContainer` - This volume mount receives all subsequent mounts
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
-Kubernetes must be configured with the OpenStack cloud provider.
-{{< /note >}}
+* `Bidirectional` - This volume mount behaves the same the `HostToContainer` mount.
+  In addition, all volume mounts created by the container will be propagated
```
```suggestion
  In addition, all volume mounts created by the container are propagated
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
+  {{< warning >}}
+  `Bidirectional` mount propagation can be dangerous. It can damage
+  the host operating system and therefore it is allowed only in privileged  
+  containers. Familiarity with Linux kernel behavior is strongly recommend  ed.
```
```suggestion
   containers. Familiarity with Linux kernel behavior is strongly recommended.
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
+  the host operating system and therefore it is allowed only in privileged  
+  containers. Familiarity with Linux kernel behavior is strongly recommend  ed.
+  In addition, any volume mounts created by containers in pods must be destroyed
+  (unmounted) by the Containers on termination.
```
```suggestion
  (unmounted) by the containers on termination.
```
?

(Also: I'm not sure what this sentence aims to convey).

geoffcline replied:
I just removed that unclear sentence, I think the general warning is sufficent.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
+  (unmounted) by the Containers on termination.
+  {{< /warning >}}
+
+### Mount Propagation Configuration
```
```suggestion
### Mount propagation configuration
```
Could this move to a new task page? If someone is new to Kubernetes, they are going to see a lot of _very_ detailed information before they have even understood what Kubernetes means by a new volume. I'm genuinely worried about driving readers away if there's too much detail too early.

geoffcline replied:
What do you think about mount propagation as a separate concept page under storage?

I think the config section is too small to survive on it's own under tasks.

sftim commented on content/en/docs/concepts/storage/volumes.md:
````diff
@@ -11,221 +11,289 @@ weight: 10
+sudo systemctl restart docker
+```
+
+## Specifiy subPaths on Volumes {#using-subpath}
````
```suggestion
## Specify subPaths on volumes {#using-subpath}
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
+
+Sometimes, it is useful to share one volume for multiple uses in a single pod. The `volumeMounts.subPath` property specifies a sub-path inside the referenced volume instead of its root.
+
+The following example shows how to configure a Pod with a LAMP stack (Linux Apache MySQL PHP)
+using a single, shared volume. This sample `subPath` configuration is not recommended
+for production use.
+
+The PHP application's code and assets map to the volume's `html` folder and
+the MySQL database is stored in the volume's `mysql` folder. For example:
```
This also feels like a task more than a concept.

geoffcline replied:
I'll move "Using subPath with expanded environment variables" out, but I think it's worth keeping the general discussion of subpaths here, as it's a simple and relevant feature.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
+    - name: mysql
+      image: mysql
+      env:
+      - name: MYSQL_ROOT_PASSWORD
+        value: "rootpasswd"
```
This needs a note about Secret management - see https://kubernetes.io/docs/tasks/run-application/run-single-instance-stateful-application/ for an example.

geoffcline replied:
added ✅

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
 ### configMap
 
-A [ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/)
+A [ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/) 
```
BTW this could link to https://kubernetes.io/docs/concepts/configuration/configmap/

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -942,358 +808,66 @@ Simultaneous writers are not allowed.
 
-The StorageOS Container requires 64-bit Linux and has no additional dependencies.
-A free developer license is available.
+## Volume Plugin Interfaces
```
```suggestion
## Storage plugin interfaces
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -942,358 +808,66 @@ Simultaneous writers are not allowed.
-
-A `vsphereVolume` is used to mount a vSphere VMDK volume into your Pod.  The contents
-of a volume are preserved when it is unmounted. It supports both VMFS and VSAN datastore.
+CSI drivers may not be compatible across all Kubernetes releases.
+Please check the specific CSI driver's documentation for supported
+deployments steps for each Kubernetes release and a compatibility matrix.
```
This looks importamt; I would wrap this in a ` {{< note >}}` &hellip;  `{{< note >}}`

geoffcline replied:
It looks like it is to me - unless you are referring to something besides the CSI drivers.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -942,358 +808,66 @@ Simultaneous writers are not allowed.
 
-{{< note >}}
-You must create vSphere VMDK volume using one of the following methods before using with a Pod.
+Support for CSI spec versions 0.2 and 0.3 are deprecated in Kubernetes
+v1.13 and will be removed in a future release.
```
We don't need to keep this.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -942,358 +808,66 @@ Simultaneous writers are not allowed.
 {{< /note >}}
 
-#### Creating a VMDK volume {#creating-vmdk-volume}
+[Container Storage Interface](https://github.com/container-storage-interface/spec/blob/master/spec.md) (CSI) defines a standard interface for container orchestration systems (like Kubernetes) to expose arbitrary storage systems to their container workloads.
```
```suggestion
The {{< glossary_tooltip text="Container Storage Interface" term_id="csi" >}}defines a standard interface for container orchestration systems (like Kubernetes) to expose arbitrary storage systems to their container workloads.
```
(it's OK to amend that glossary entry to, eg, add a further hyperlink).

geoffcline replied:
I added the link to the next paragraph.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1391,72 +963,489 @@ plugin path on each node and in some cases the control plane nodes as well.
 
-Mount propagation of a volume is controlled by the `mountPropagation` field
-in `Container.volumeMounts`. Its values are:
+See the information about [PersistentVolumes](/docs/concepts/storage/persistent-volumes/) for more
+details.
```
Also consider linking to PersistentVolumeClaim.

geoffcline replied:
the introduction of the linked page introduces both terms

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1391,72 +963,489 @@ plugin path on each node and in some cases the control plane nodes as well.
-  that are mounted to this volume or any of its subdirectories by the host.
-  In similar fashion, no mounts created by the container will be visible on
-  the host. This is the default mode.
+## External Volume Providers
```
```suggestion
## External storage providers
```
I'm not sure that I could explain what these are external to.

geoffcline replied:
The concept was "externally managed" (i.e., cloud providers, X as a service), but I'm open to alternative terminology.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1391,72 +963,489 @@ plugin path on each node and in some cases the control plane nodes as well.
 
-* `HostToContainer` - This volume mount will receive all subsequent mounts
-  that are mounted to this volume or any of its subdirectories.
+An `awsElasticBlockStore` volume mounts an Amazon Web Services (AWS) [EBS
```
```suggestion
An `awsElasticBlockStore` volume mounts a filesystem stored on an Amazon
Web Services (AWS) _[EBS](https://aws.amazon.com/ebs/) volume_ into your pod.
Unlike
```
?
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1391,72 +963,489 @@ plugin path on each node and in some cases the control plane nodes as well.
-  In other words, if the host mounts anything inside the volume mount, the
-  container will see it mounted there.
+{{< caution >}}
+You must create an EBS volume using `aws ec2 create-volume` or the AWS API before you can use it.
```
BTW, Kubernetes now offers [dynamic provisioning](https://kubernetes.io/docs/concepts/storage/dynamic-provisioning/) for storage.

geoffcline replied:
noted, thanks!

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1391,72 +963,489 @@ plugin path on each node and in some cases the control plane nodes as well.
-  In addition, any volume mounts created by containers in pods must be destroyed
-  (unmounted) by the containers on termination.
-  {{< /warning >}}
+Make sure the zone matches the zone you brought up your cluster in.  (And also check that the size and EBS volume
```
```suggestion
Make sure the zone matches the zone where you want to mount that volume.
(And also check that the size and EBS volume
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
````diff
@@ -1391,72 +963,489 @@ plugin path on each node and in some cases the control plane nodes as well.
-```
 
 ## {{% heading "whatsnext" %}}
 
````
I would direct the reader to read https://k8s.io/docs/concepts/storage/persistent-volumes/

sftim reviewed (commented):
Hi @geoffcline

Here's some early feedback.
Don't worry if you don't agree with everything mentioned here or if this seems like a lot to fix. This concept page is in need of updating and it might take more than one pull request before it's current. In fact, it's probably better to make incremental small changes. Feel free to highlight which areas you're tackling so that reviewers can skip feedback about any shortcomings you're saving for some future PR.

geoffcline commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
-#### AWS EBS configuration example
+Volumes can not mount into other volumes, or have hard links to
+other volumes. Each container in the pod's configuration must independently specify where to
+mount each volume. 
```
given that this is an example, I think it's fair to have the example stay in the scope of filesystem volumes. I clarified this at the start of the example.

geoffcline commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
+A Kubernetes volume's lifecycle is tied to the *pod* that encloses it. The pod maintains the volume as different containers are started or stopped. Multiple volumes of any type may be connected to a pod.
 
 At its core, a volume is just a directory, possibly with some data in it, which
-is accessible to the containers in a pod. How that directory comes to be, the
+is accessible to the containers in a pod.  How that directory comes to be, the
 medium that backs it, and the contents of it are determined by the particular
 volume type used.
```
I added a brief mention of this. I'm considering opening an issue to add docs for volumeMode Block

sftim replied:
“At its core, a volume is just a directory” is incorrect in the original. An issue about cleaning that up sounds like a good idea.

geoffcline replied:
I changed this to "One use case for volumes is providing directories to pods, possibly with some data in it."

geoffcline commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,289 @@ weight: 10
-must be installed on the cluster and the `CSIMigration` and `CSIMigrationAWS`
-beta features must be enabled.
+Core Volume Types:
+   * [cephfs](#cephfs)
+   * [configMap](#configmap)
+   * [downwardAPI](#downwardapi)
+   * [emptyDir](#emptydir)
+   * [fc (fibre channel)](#fc)
+   * [gitRepo (deprecated)](#gitrepo)
+   * [glusterfs](#glusterfs)
+   * [hostPath](#hostpath)
+   * [iscsi](#iscsi)
+   * [local](#local)
+   * [nfs](#nfs)
+   * [projected](#projected)
+   * [rbd](#rbd)
+   * [secret](#secret)
```
how about I move cephfs, gitRepo, glusterfs, rbd to the other category

and keep iscsi and nfs, as they are long standing generic network storage protocols

sftim replied:
iSCSI and NFS both have CSI drivers so I'd include those in the “tied to particular technologies” grouping.

tengqm commented:
Do we have a clear goal for this page revision? It is over 1300 lines and we are revising the text everywhere.
The implication to localization teams is that they need to retranslate the whole page.

geoffcline commented:
I'm partially responding to this issue https://github.com/kubernetes/website/issues/24726

Also, my own research into the docs site analytics indicated this volumes page is popular, and also isn't meeting users needs. 

The goal is driven by user experience. The current page is confusing and unhelpful. The revised page has an example and general information up front, then a more organized listing of specific volume types. 

I agree this is a large change, but that's necessary when reorganizing a page. 

Respectfully, the large change is a worthwhile improvement.

geoffcline commented:
link to draft page

https://deploy-preview-25042--kubernetes-io-master-staging.netlify.app/docs/concepts/storage/volumes/

geoffcline commented:
thread from wg-naming re new category names

https://kubernetes.slack.com/archives/C016DM451K4/p1599507620004800

kbhawkey commented:
Hi @geoffcline . I read through the changes. Here is some initial feedback:
- There are a number of minor formatting/spelling/heading text changes to correct.
- The "Background" paragraphs cover a range of topics (including a basic volume configuration sample).
- Do the new categories help organize the content and make sense for readers learning about storage options?

Proposed main sections:
Core volume types provided by Kubernetes
Storage plugin interfaces
Volume types not provided by Kubernetes but interface with Kubernetes (CSI)

I suggest to get feedback from SIG Storage.
/cc @kubernetes/sig-storage-pr-reviews

geoffcline commented:
Thank you for your feedback. I've updated the title.

> The "Background" paragraphs cover a range of topics (including a basic volume configuration sample).

I've changed this heading to "introduction" to be clearer. Overall, I think the intro paragraphs make this page more useful. 

> re: breakdown and SIG storage.

I asked a while ago in their slack channel about the breakdown, and I tried to base the breakdown on that discussion. Hopefully we can get more feedback on this PR. 

👍

kbhawkey commented on content/en/docs/concepts/storage/mount-propagation.md:
```diff
@@ -0,0 +1,77 @@
+weight: 40
+---
+
+<!-- overview -->
```
How does this content fit with the updates to the Volumes page?

geoffcline replied:
I responded to previous feedback that this content was too narrow/technical to include in the volumes page, and should be moved to a separate task. I'm flexible on where to put it.

However, I would like to keep it in the same PR otherwise it's likely to be dropped completely.

kbhawkey commented on content/en/docs/tasks/administer-cluster/subpath-env.md:
```diff
@@ -0,0 +1,54 @@
+---
+title: Using subPath with expanded environment variables 
+content_type: task
+---
```
Is this task page related to the updates to the Volumes page?

geoffcline replied:
I think it's related to the concept of "refactoring" the volumes page. In response to feedback, I broke those out into separate pages.

geoffcline commented:
I pushed a revision that addressed some of the typos/grammar issues. ✔

sftim reviewed (commented):
Hi @geoffcline 

Could you squash (possibly also rebase) this down to a small number of commits, perhaps 1 or 2? That will make this PR easier to review.

If you'd like help with that squash, I can try to make time to explain what's involved. The process does involve a bit of Git-specific knowledge.

sftim commented on content/en/docs/concepts/storage/mount-propagation.md:
```diff
@@ -0,0 +1,77 @@
+
+<!-- overview -->
+
+Mount propagation allows for sharing volumes mounted by a container to
+other containers in the same pod, or even to other pods on the same node.
```
Consider using glossary tooltips here for at least volume, and possibly also container, pod and node.

geoffcline replied:
✔

sftim commented on content/en/docs/concepts/storage/mount-propagation.md:
```diff
@@ -0,0 +1,77 @@
+   containers. Familiarity with Linux kernel behavior is strongly recommended.
+  {{< /warning >}}
+
+### Mount Propagation Configuration
```
nit: Subheadings should normally be sentence case.

sftim commented on content/en/docs/concepts/storage/mount-propagation.md:
````diff
@@ -0,0 +1,77 @@
+  {{< /warning >}}
+
+### Mount Propagation Configuration
+Before mount propagation can work properly on some deployments (CoreOS,
+RedHat/Centos, Ubuntu) mount share must be configured correctly in
+Docker as shown below.
+
+Edit your Docker's `systemd` service file. Set `MountFlags` as follows:
+```shell
+MountFlags=shared
+```
````
Is Docker a prerequisite?

sftim commented on content/en/docs/tasks/administer-cluster/subpath-env.md:
```diff
@@ -0,0 +1,54 @@
+---
+title: Using subPath with expanded environment variables 
```
nit: page titles should normally be in title case

sftim commented on content/en/docs/tasks/administer-cluster/subpath-env.md:
```diff
@@ -0,0 +1,54 @@
+---
+title: Using subPath with expanded environment variables 
+content_type: task
+---
```
```suggestion
min-kubernetes-server-version: v1.17
---
```
(is that right?)

geoffcline replied:
I can find it in 1.15, so I'll use that.

sftim commented on content/en/docs/tasks/administer-cluster/subpath-env.md:
```diff
@@ -0,0 +1,54 @@
+{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}
+
+<!-- steps -->
+
```
This jumps into a manifest with little context. What task is the reader learning to perform?

geoffcline replied:
I have revised this section with a new, friendlier, intro.

sftim commented on content/en/docs/tasks/administer-cluster/subpath-env.md:
```diff
@@ -0,0 +1,54 @@
+
+<!-- steps -->
+
+## Example Spec
```
nit: This example is actually a manifest

sftim commented on content/en/docs/concepts/storage/mount-propagation.md:
```diff
@@ -0,0 +1,77 @@
+
+## {{% heading "whatsnext" %}}
+
+Learn more about managing storage declaratively using [Persistent Volumes](https://k8s.io/docs/concepts/storage/persistent-volumes/).
```
nit:
```suggestion
Learn more about managing storage declaratively using [Persistent Volumes](/docs/concepts/storage/persistent-volumes/).
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,196 @@ weight: 10
-The Kubernetes {{< glossary_tooltip text="volume" term_id="volume" >}} abstraction
-solves both of these problems.
-Familiarity with [Pods](/docs/concepts/workloads/pods/) is suggested.
+File storage is useful when building stateful applications. Use _volumes_ to declaratively manage storage for containers. The Kubernetes {{< glossary_tooltip text="volume" term_id="volume" >}} abstraction allows files to persist even beyond the lifetime of a pod. Volumes also allow different containers within a pod to share storage and to access configuration.
```
File storage is sometimes also useful when building stateless applications. For example, an `emptyDir` volume can be used for a cache, or a Pod might mount service account data via a volume.
I recommend rewording to avoid implying that only stateful apps would use volumes.

geoffcline replied:
good point

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,196 @@ weight: 10
-Familiarity with [Pods](/docs/concepts/workloads/pods/) is suggested.
+File storage is useful when building stateful applications. Use _volumes_ to declaratively manage storage for containers. The Kubernetes {{< glossary_tooltip text="volume" term_id="volume" >}} abstraction allows files to persist even beyond the lifetime of a pod. Volumes also allow different containers within a pod to share storage and to access configuration.
+
+Familiarity with [Pods](/docs/concepts/workloads/pods/pod/) is suggested.
```
nit:
```suggestion
Familiarity with [Pods](/docs/concepts/workloads/pods/) is suggested.
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,196 @@ weight: 10
-the lifetime of a pod. Consequently, a volume outlives any containers
-that run within the pod, and data is preserved across container restarts. When a
-pod ceases to exist, the volume is destroyed.
+A Kubernetes volume's lifecycle is tied to the *pod* that encloses it. The pod maintains the volume as different containers are started or stopped. Multiple volumes of any type may be connected to a pod. The [PersistentVolume](/docs/concepts/storage/persistent-volumes/) abstraction builds on the volume concept to provide storage that outlasts the lifetime of a pod.
```
How about:
```suggestion
A Kubernetes volume's lifecycle is tied to the *pod* that encloses it. The pod maintains the volume as different containers are started or stopped. Multiple volumes of any type may be connected to a pod. The [PersistentVolume](/docs/concepts/storage/persistent-volumes/) abstraction builds on the volume concept to let pods access storage that outlasts the lifetime of a pod.
```

geoffcline replied:
I was attempting to avoid passive voice, but in this case it's the best option.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,196 @@ weight: 10
-unmounted. This means that a `cephfs` volume can be pre-populated with data, and
-that data can be shared between pods. The `cephfs` volume can be mounted by multiple
-writers simultaneously.
+The Kubernetes binary includes some volume types for communicating with externally managed volumes, such as AWS EBS or Azure Disk Storage. These volume types are described as in-tree storage plugins. 
```
There is no `kubernetes` binary. Maybe reword?

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,196 @@ weight: 10
-{{< note >}}
-Kubernetes must be configured with the OpenStack cloud provider.
-{{< /note >}}
+Sometimes, it is useful to share one volume for multiple uses in a single pod. The `volumeMounts.subPath` property specifies a sub-path inside the referenced volume instead of its root.
```
`volumeMounts.subPath` can be useful even if the volume is mounted just once in a Pod. For example, if the volume layout is determined externally from Kubernetes.

In this example we mount the same volume twice, and that's useful to show that it's possible.

geoffcline replied:
I reworded this to indicate the multiple mounts in a pod is an example.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,196 @@ weight: 10
+Note: The password is defined in the config yaml, which is insecure. See Kubernetes Secrets for a secure solution.
 
-{{< feature-state for_k8s_version="v1.18" state="beta" >}}
+## Core Volume Types
```
```suggestion
## Built-in volume types
```

The full list of built-in volume types could move to the reference section; for example, in a follow-up PR.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,221 +11,196 @@ weight: 10
+before you can use it.
+
+{{< note >}}
+A container using a ConfigMap as a [subPath](#using-subpath) volume mount will not receive ConfigMap updates.
```
nit:
```suggestion
A container using a ConfigMap as a [subPath](#using-subpath) volume mount does not receive ConfigMap updates.
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -249,43 +224,34 @@ spec:
-
-* Text data is exposed as files using the UTF-8 character encoding. For other character encodings, use `binaryData`.
-{{< /note >}}
+Text data is exposed as files using UTF-8 character encoding. To use some other character encoding, use binaryData.
```
```suggestion
Values from ConfigMaps are exposed as text files, using UTF-8 character encoding. To use some other character encoding, or to store information that is not text, use the `binaryData` (rather than `data`) field in the ConfigMap.
```
Something like that? Could save rewording this for another PR, BTW.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -249,43 +224,34 @@ spec:
 {{< note >}}
-A container using the downward API as a [`subPath`](#using-subpath) volume mount will not
-receive downward API updates.
+A container using downward API as a [`subPath`](#using-subpath) volume mount will not receive downward API updates.
```
Why this change? The original text makes more sense to me.

geoffcline replied:
not sure why I changed that -- reverted ✔

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -541,29 +371,42 @@ The supported values for field `type` are:
+`hostPath` volume can consume, and no isolation between containers or between
+pods.
+
+In the future, we expect that `emptyDir` and `hostPath` volumes will be able to
+request a certain amount of space using a [resource](/docs/concepts/configuration/manage-resources-containers/)
+specification, and to select the type of media to use, for clusters that have
+several media types.
```
We usually [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future).

geoffcline replied:
I've revised this section with more updated information, based on https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#setting-requests-and-limits-for-local-ephemeral-storage

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -644,8 +484,8 @@ partition or directory.
 
-Compared to `hostPath` volumes, `local` volumes are used in a durable and
-portable manner without manually scheduling pods to nodes. The system is aware
+Compared to `hostPath` volumes, local volumes are used in a *durable and
```
The original is correct to write `local`, I think. If `hostPath` is in backticks then `local` should be too.
```suggestion
Compared to `hostPath` volumes, `local` volumes are used in a durable and
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -912,86 +702,23 @@ A container using a projected volume source as a [`subPath`](#using-subpath) vol
-an `rbd` volume are preserved and the volume is unmounted. This
-means that a RBD volume can be pre-populated with data, and that data can
-be shared between pods.
+An `rbd` volume allows a[Rados Block Device](https://ceph.com/docs/master/rbd/rbd/) (RBD) volume to mount into your pod. Unlike `emptyDir`, which is erased when a pod is removed, the contents of an `rbd` volume are preserved and the volume is unmounted. This means that a RBD volume can be pre-populated with data, and that data can be shared between pods. 
```
nit:
```suggestion
An `rbd` volume allows a [Rados Block Device](https://ceph.com/docs/master/rbd/rbd/) (RBD) volume to mount into your pod. Unlike `emptyDir`, which is erased when a pod is removed, the contents of an `rbd` volume are preserved and the volume is unmounted. This means that an RBD volume can be pre-populated with data, and that data can be shared between pods.
```
(this text'll be easier to work with if wrapped, though)

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1000,469 +727,627 @@ backed by tmpfs (a RAM-backed filesystem) so they are never written to
+provisioning/delete, attach/detach, mount/unmount and resizing of volumes.
 
-#### vSphere CSI migration complete {#vsphere-csi-migration-complete}
+In-tree plugins that support `CSIMigration` and have a corresponding CSI driver implemented are listed in [Types of Volumes](#volume-types).
```
If an in-tree plugin doesn't support CSI migration it's listed separately? Why this distinction?

geoffcline replied:
I'll rephrase this, it's really just moving things that are superseded by a CSI approach to a separate section.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1000,469 +727,627 @@ backed by tmpfs (a RAM-backed filesystem) so they are never written to
 kind: Pod
 metadata:
-  name: my-lamp-site
+  name: test-ebs
```
`test-ebs` doesn't feel like a good name for a Pod, to be honest.

geoffcline replied:
how about `demo-webserver`?

geoffcline replied:
I used "webserver-using-aws-ebs" based on the other comment.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1000,469 +727,627 @@ backed by tmpfs (a RAM-backed filesystem) so they are never written to
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test-pd
```
How about `webserver-using-gce-pd`?

sftim reviewed (commented):
Hi. Here's some more feedback.

I recommend focusing on the review items that are about page structure and content accuracy ahead of things like typos. We can always log a separate issue to clean up each typo, but page structure does _not_ lend itself to good-first-issue tasks.

geoffcline commented:
Ok!

✔Rebased & squashed
✔Responded to @sftim 's comments
✔Pinged sig-storage to review category breakdown

Thanks again for the feedback

kbhawkey commented:
👋 @geoffcline .
Would you rebase your changes?
Pinging Sig Storage for review:
/cc @kubernetes/sig-storage-pr-reviews 


Page previews:
https://deploy-preview-25042--kubernetes-io-master-staging.netlify.app/docs/concepts/storage/volumes/
https://deploy-preview-25042--kubernetes-io-master-staging.netlify.app/docs/concepts/storage/mount-propagation/
https://deploy-preview-25042--kubernetes-io-master-staging.netlify.app/docs/tasks/administer-cluster/subpath-env/

geoffcline commented:
Thanks @kbhawkey  👍 I pushed a rebase.

sftim commented on content/en/docs/concepts/storage/mount-propagation.md:
```diff
@@ -0,0 +1,77 @@
+
+<!-- overview -->
+
+Mount propagation allows for sharing a {{< glossary_tooltip text="cluster" term_id="volume" >}} mounted by a container to
+other containers in the same {{< glossary_tooltip text="cluster" term_id="pod" >}}, or even to other pods on the same {{< glossary_tooltip text="cluster" term_id="node" >}}.
```
```suggestion
Mount propagation allows for sharing a {{< glossary_tooltip text="volume" term_id="volume" >}} mounted
by a container to other containers in the same {{< glossary_tooltip text="pod" term_id="pod" >}}, or even to other pods on the same {{< glossary_tooltip text="node" term_id="node" >}}.
```

sftim commented on content/en/docs/tasks/administer-cluster/subpath-env.md:
```diff
@@ -0,0 +1,63 @@
+In this example, a pod uses `subPathExpr` to create a directory `pod1` within
+the `hostPath` volume `/var/log/pods`. All containers see this as simply `/logs`. Thus, logs may be collected from multiple pods in directories such as `/var/log/pods/pod1/` and `.../pods/pod2/`.
+
+First, the spec defines an environment variable `POD_NAME` using the `downwardAPI`.
```
I'd write:
```suggestion
First, the spec defines an environment variable `POD_NAME` using the _downward API_.
```

sftim commented on content/en/docs/tasks/administer-cluster/subpath-env.md:
```diff
@@ -0,0 +1,63 @@
+
+## Example manifest
+
+In this example, a pod uses `subPathExpr` to create a directory `pod1` within
+the `hostPath` volume `/var/log/pods`. All containers see this as simply `/logs`. Thus, logs may be collected from multiple pods in directories such as `/var/log/pods/pod1/` and `.../pods/pod2/`.
```
Should we warn people that read-write `hostPath` mounts are usually not suited for production (and that [`local`](https://kubernetes.io/docs/concepts/storage/volumes/#local) might be a better option? The original page includes that warning, albeit quite a long way away from this text.

(Perhaps we could use a `local` volume in the example - but that sounds like a separate fix).

geoffcline replied:
I think i'll open an issue to make better examples using local.

geoffcline replied:
done! #27439

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1002,469 +720,629 @@ backed by tmpfs (a RAM-backed filesystem) so they are never written to
 
-  This mode is equal to `rshared` mount propagation as described in the
-  [Linux kernel documentation](https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt)
+A `vsphereVolume` is used to mount a vSphere VMDK Volume into your pod.  The contents of a volume are preserved when it is unmounted. It supports both VMFS and VSAN datastore.
```
```suggestion
A `vsphereVolume` is used to mount a VMWare vSphere VMDK Volume into your pod.  The contents of a volume are preserved when it is unmounted. It supports both VMFS and VSAN datastore.
```
?

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1002,469 +720,629 @@ backed by tmpfs (a RAM-backed filesystem) so they are never written to
+beta features must be enabled.
 
-## Out-of-tree volume plugins
+### flocker (depreceated) {#flocker}
```
```suggestion
### flocker (deprecated) {#flocker}
```

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1002,469 +720,629 @@ backed by tmpfs (a RAM-backed filesystem) so they are never written to
-You can set up your
-[PersistentVolume/PersistentVolumeClaim with raw block volume support](/docs/concepts/storage/persistent-volumes/#raw-block-volume-support) as usual, without any CSI specific changes.
+#### Manually provisioning a Regional PD PersistentVolume
+Dynamic provisioning is possible using a [StorageClass for GCE PD](/docs/concepts/storage/storage-classes/#gce).
```
The correct link is actually like this:
```suggestion
Dynamic provisioning is possible using a [StorageClass for GCE PD](/docs/concepts/storage/storage-classes/#gce-pd).
```

sftim reviewed (commented):
This is looking good.

I've suggested a few tweaks.

geoffcline commented:
Ok after much delay (sorry), I've updated this.

Thanks for the fixes and catching those incorrect labels Tim :)

I decided to keep the segments of volume types the same. I didn't hear much consensus around an alternate way. Also, I think as CSI becomes more mature, the plugins superseded by CSI will be documented under that directly, or have docs elsewhere.

sftim commented:
Informally, LGTM

geoffcline commented:
I have done the necessary rebase.

I understand we have some differences on how to split up the volume types into sections. I've reached out to the storage sig and they didn't seem to have any strong opinions/suggestions. I think separating the volume types with a CSI migration makes some sense, as the relevant docs will (eventually) fall under the CSI sections more directly. 

Let's get this merged! :)

kbhawkey commented on content/en/docs/tasks/administer-cluster/subpath-env.md:
```diff
@@ -0,0 +1,71 @@
+
+* Learn more about [PersistentVolumes](/docs/concepts/storage/persistent-volumes/).
+<<<<<<< HEAD
+<<<<<<< HEAD
```
Hi @geoffcline .
Would you check these additions and clean up any rebasing text?

geoffcline replied:
thanks!

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
+A Kubernetes volume's lifecycle is tied to the *pod* that encloses it. The pod maintains the volume as different containers are started or stopped. Multiple volumes of any type may be connected to a pod. The [PersistentVolume](/docs/concepts/storage/persistent-volumes/) abstraction builds on the volume concept to let pods access storage that outlasts the lifetime of a pod.
 
-Before you can use an EBS volume with a pod, you need to create it.
+One use case for volumes is providing directories to pods, possibly with some data in it.  How that directory comes to be, the medium that backs it, and the contents of it are determined by the particular volume type used.
```
I think the statement about "providing a filesystem to pods" is pretty basic and should come before the lifecycle part.

geoffcline replied:
Ok, I reorganized some paragraphs to address this.

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
 
-Make sure the zone matches the zone you brought up your cluster in. Check that the size and EBS volume
-type are suitable for your use.
+In addition to filesystem volumes, Kubernetes also supports volumes in a raw block access mode. For example, The `volumeMode` field can be set to `block` in the specifications for PersistentVolume and PersistentVolumeClaim. 
```
"Kubernetes also support persistent volumes"

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
-type are suitable for your use.
+In addition to filesystem volumes, Kubernetes also supports volumes in a raw block access mode. For example, The `volumeMode` field can be set to `block` in the specifications for PersistentVolume and PersistentVolumeClaim. 
+ 
+### Volume specification example
```
This seems too early. We don't know what an emptydir volume is yet.

geoffcline replied:
I think it's important to have an example early on, people will usually scroll to it. I added additional description and a link to emptydir.

msau42 commented on content/en/docs/concepts/storage/volumes.md:
````diff
@@ -11,228 +11,193 @@ weight: 10
+```
 
-For more details, see the [`azureDisk` volume plugin](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/staging/volumes/azure_disk/README.md).
+### Core volume types {#volume-types-core}
````
I think it may make more sense to divide the volume types by "kubernetes core" and "general purpose storage".

Kubernetes core are specialized volume types (mostly ephemeral) and are not used for general purpose persistent storage use cases. They include:
- configmap
- downwardapi
- emptydir
- gitrepo
- hostpath
- projected
- secret

The remaining types are all general purpose storage and most commonly used as PersistentVolumes

geoffcline replied:
Ok, I have been spinning my wheels on how to break this down. I think that "Kubernetes core" (i.e., Kubernetes specific), and "general purpose" are better categories. 

I had previously considered "externally hosted" and "native" but there were a number of pitfalls. However, I think "Kubernetes specific" will work well.

geoffcline replied:
I have implemented this change. I sincerely appreciate your insight on this particular issue, it has resolved a persistent (ha) problem I had with this PR.

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
 
-The `azureFile` volume type mounts a Microsoft Azure File volume (SMB 2.1 and 3.0)
-into a pod.
+A [persistentVolumeClaim](#persistentvolumeclaim) is a declarative way to ask for storage, without preference for the underlying implementation. 
```
The most important thing about PVCs is that their lifetime is independent of a pod. So data will persist beyond a pod's lifetime.

geoffcline replied:
I think a reader will understand that from the title and the basic introduction. The "asking for storage without preference" is what makes PVCs noteworthy.

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
+A [persistentVolumeClaim](#persistentvolumeclaim) is a declarative way to ask for storage, without preference for the underlying implementation. 
 
-For more details, see the [`azureFile` volume plugin](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/staging/volumes/azure_file/README.md).
+This is powered by storage plugins. More specifically, Kubernetes implements the {{< glossary_tooltip text="Container Storage Interface" term_id="csi" >}} (CSI),
```
Unfortunately at this stage, we still have "in-tree" plugins and "csi" plugins. They both support PVCs so I don't think we should only talk about CSI here. Or maybe we name this section "Storage Extensions". But then I would remove the part about PVCs.

geoffcline replied:
I added a short paragraph addressing this.

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
+to learn more about CSI. 
 
-#### azureFile CSI migration
+[FlexVolume](#flexvolume) is an another mechanism for storage plugins. Unlike CSI, FlexVolume is
```
We should say this is deprecated and CSI is the recommended way to develop storage plugins for Kubernetes.

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
+   * [azureDisk](#azuredisk)
+   * [azureFile](#azurefile)
+   * [cinder](#cinder)
+   * [flocker](#flocker)
```
deprecated

geoffcline replied:
just to be clear -- should I delete the documentation for these deprecated types?

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
-{{< note >}}
-You must have your own Ceph server running with the share exported before you can use it.
-{{< /note >}}
+The CSI was introduced to standardize connectivity to externally managed volumes, such as cloud platform storage. However, Kubernetes still includes many volume types for connecting to specific cloud storage providers. 
```
Reword something to something like "Using in-tree plugins are not recommended anymore. All new storage features are only supported for CSI volumes, and in-tree plugins are being replaced by their equivalent CSI drivers".

geoffcline replied:
added a callout on this

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
+   * [flocker](#flocker)
+   * [gcePersistentDisk](#gcepersistentdisk)
+   * [portworxVolume](#portworxvolume)
+   * [quobyte](#quobyte)
```
deprecated

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
+   * [gcePersistentDisk](#gcepersistentdisk)
+   * [portworxVolume](#portworxvolume)
+   * [quobyte](#quobyte)
+   * [storageos](#storageos)
```
deprecated

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
+   * [portworxVolume](#portworxvolume)
+   * [quobyte](#quobyte)
+   * [storageos](#storageos)
+   * [scaleIO](#scaleio)
```
deprecated

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
-{{< note >}}
-Kubernetes must be configured with the OpenStack cloud provider.
-{{< /note >}}
+The `volumeMounts.subPath` property specifies a sub-path inside the referenced volume instead of its root. For example, this property may be used to to share one volume for multiple uses in a single pod.
```
"The `volumeMounts.subPath` property mounts the subdirectory of the volume into the container"

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
+   * [vsphereVolume](#vspherevolume)
 
-### cinder
+## Specify subPaths on volumes {#using-subpath}
```
The ordering of this section seems out of place? It's right in the middle of documenting the volume types

geoffcline replied:
I moved it up to after the spec example.

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -338,71 +294,107 @@ spec:
+
+#### Resources
+
+The storage media (such as Disk or SSD) of an `emptyDir` volume is determined by the medium of the filesystem holding the kubelet root dir (typically
```
Move the references to `emptyDir` to the emptydir section?

geoffcline replied:
done, thank you for noticing this

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1007,300 +725,47 @@ backed by tmpfs (a RAM-backed filesystem) so they are never written to
 
-A `storageos` volume allows an existing [StorageOS](https://www.storageos.com)
-volume to mount into your Pod.
+## Storage plugin interfaces
```
The "storage plugin" section earlier in this doc seems to duplicate some of this section?

geoffcline replied:
I intended that earlier section to be a quick overview of the options, and this section to be a bit more detailed.

msau42 replied:
If this section is only for out of tree plugin extensions should we call this something like "External plugin extensions"

msau42 replied:
In terms of ordering, should we put this before or after listing all the in-tree types? fit seems a little strange that we're talking about external plugins in between 2 in-tree plugin sections

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1404,72 +867,493 @@ plugin path on each node and in some cases the control plane nodes as well.
 
-Mount propagation of a volume is controlled by the `mountPropagation` field
-in `Container.volumeMounts`. Its values are:
+A `persistentVolumeClaim` volume is used to mount a
```
I think the most imprtant aspect of a PVC is that the lifecycle is independent of a pod

geoffcline replied:
I added a note calling this out, thanks

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -1404,72 +867,493 @@ plugin path on each node and in some cases the control plane nodes as well.
 ## {{% heading "whatsnext" %}}
 
-Follow an example of [deploying WordPress and MySQL with Persistent Volumes](/docs/tutorials/stateful-application/mysql-wordpress-persistent-volume/).
+Learn more about managing storage declaratively using [Persistent Volumes](https://k8s.io/docs/concepts/storage/persistent-volumes/).
```
I think the most important part is the persistent lifecycle, not the declaritive nature

geoffcline replied:
I reworded this.

msau42 commented on content/en/docs/tasks/administer-cluster/subpath-env.md:
```diff
@@ -0,0 +1,71 @@
+
+## {{% heading "whatsnext" %}}
+
+* Learn more about [PersistentVolumes](/docs/concepts/storage/persistent-volumes/).
```
I would probably point to the volumes page as well

geoffcline replied:
done!

msau42 commented on content/en/docs/concepts/storage/mount-propagation.md:
```diff
@@ -0,0 +1,77 @@
+
+## {{% heading "whatsnext" %}}
+
+Learn more about managing storage declaratively using [Persistent Volumes](docs/concepts/storage/persistent-volumes/).
```
I think the most important part is the persistence, not the declaritive nature

geoffcline replied:
I agree ,but I wanted to give the user a bit of context about the page. I think between the title and the additional text, it's clear. I could rephrase to 

> Learn more about Persistent Volumes and managing storage declaratively

msau42 commented on content/en/docs/concepts/storage/volumes.md:
````diff
@@ -1404,72 +867,493 @@ plugin path on each node and in some cases the control plane nodes as well.
-sudo systemctl restart docker
-```
 
 ## {{% heading "whatsnext" %}}
````
Should we point to the persistent volumes page and any of the storage tasks?

geoffcline replied:
We already link to the persistent volumes page. I'll add links to mount propagation, and sub-path ex.

geoffcline commented:
thanks for the review @msau42 and @kbhawkey ! I'm working on it now

geoffcline commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
-between containers running together in a `Pod`.
-The Kubernetes {{< glossary_tooltip text="volume" term_id="volume" >}} abstraction
-solves both of these problems.
+File storage is useful when building applications, such as to store state or cache data. Use _volumes_ to declaratively manage storage for containers. The Kubernetes {{< glossary_tooltip text="volume" term_id="volume" >}} abstraction allows files to persist even beyond the lifetime of a pod. Volumes also allow different containers within a pod to share storage and to access configuration.
```
I think this is a fair overview of what the user will learn in this page. Additionally, in context the comment on persistence makes sense.

`The Kubernetes {{< glossary_tooltip text="volume" term_id="volume" >}} abstraction allows files to persist even beyond the lifetime of a pod. `

Volumes are allowed to persist :)

geoffcline commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
-* the nodes on which pods are running must be AWS EC2 instances
-* those instances need to be in the same region and availability zone as the EBS volume
-* EBS only supports a single EC2 instance mounting a volume
+By default, everything with the filesystems of a container is ephemeral. During startup, each pod in a container gets a filesystem created based on the layers in its {{< glossary_tooltip text="container image" term_id="image" >}}. When a container is restarted or destroyed, that root filesystem, including any modifications to it, is lost.
```
Good point. I incorporated your changes. I kept the bit on building the filesystem, as context for the lifecycle.

geoffcline replied:
Here is how I'm thinking of it.

base knowledge: docker containers (ephemeral filesystem by default)
step 1. the kuberentes volume resource
step 2. the kuberentes persistent volume resource
step 3. the kuberentes ephemeral knowledge resource.

geoffcline replied:
I changed it to this, in addition to changing the surrounding paragraphs. 

> In a standard docker container, for example, the filesystem of a container is ephemeral. When a container is restarted or destroyed, modifications to the container filesystem are lost.  
> 
> A Kubernetes volume's lifecycle is tied to the *pod* that encloses it.

geoffcline commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,193 @@ weight: 10
+By default, everything with the filesystems of a container is ephemeral. During startup, each pod in a container gets a filesystem created based on the layers in its {{< glossary_tooltip text="container image" term_id="image" >}}. When a container is restarted or destroyed, that root filesystem, including any modifications to it, is lost.
 
-#### Creating an AWS EBS volume
+A Kubernetes volume's lifecycle is tied to the *pod* that encloses it. The pod maintains the volume as different containers are started or stopped. Multiple volumes of any type may be connected to a pod. The [PersistentVolume](/docs/concepts/storage/persistent-volumes/) abstraction builds on the volume concept to let pods access storage that outlasts the lifetime of a pod.
```
I need to think on this further.

geoffcline replied:
I think it's confusing how "ephemeral volumes" describes a property of some volumes, while "PersistentVolume" is a specific resource.

geoffcline replied:
I revised the "Volume Lifecycle" section to explain the differences.

geoffcline commented:
I pushed the revised copy, and I also re-wrapped the lines. Thanks!

msau42 commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -919,388 +654,56 @@ A container using a projected volume source as a [`subPath`](#using-subpath) vol
-You must have a Ceph installation running before you can use RBD.
-{{< /note >}}
+The out-of-tree volume plugins include the Container Storage Interface (CSI)
+and FlexVolume. They enable storage vendors to create custom storage plugins. The main Kubernetes binary implements the CSI, and communicates with any external plugins. This keeps the main binary efficient, and allows users to choose their own storage plugins. 
```
There's many Kubernetes binaries. Also they implement both CSI and FlexVolume interfaces.

sftim replied:
Yep, it's wrong to write “the main Kubernetes binary”. It's OK to mention having code (not that 3.8 GiB is _that_ small) in-tree and easier ways for vendors to work at their own pace.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,191 @@ weight: 10
-[feature gates](/docs/reference/command-line-tools-reference/feature-gates/) must be enabled.
 
-Azure File CSI driver does not support using same volume with different fsgroups, if Azurefile CSI migration is enabled, using same volume with different fsgroups won't be supported at all.
+### General purpose volume types
```
Sounds reasonable
I'd clarify below the header, I think.

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,191 @@ weight: 10
-unmounted. This means that a `cephfs` volume can be pre-populated with data, and
-that data can be shared between pods. The `cephfs` volume can be mounted by multiple
-writers simultaneously.
+The CSI was introduced to standardize connectivity to externally managed volumes, such as cloud platform storage. However, Kubernetes still includes many volume types for connecting to specific cloud storage providers. 
```
Is CSI useful outside Kubernetes (like CNI and CRI are)?

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,191 @@ weight: 10
+   * [gitRepo (deprecated)](#gitrepo)
+   * [glusterfs](#glusterfs)
+   * [hostPath](#hostpath)
+   * [local](#local)
```
How come `local` and `hostPath` get separated? I'm OK to keep them in separate sections but we should have a good reason for that (probably a good enough reason that we raise an issue about mentioning it in the docs!).

sftim commented on content/en/docs/concepts/storage/volumes.md:
```diff
@@ -11,228 +11,191 @@ weight: 10
-{{< feature-state for_k8s_version="v1.21" state="beta" >}}
+   * [CSI](#csi)
+   * [FlexVolume](#flexvolume)
+   * [persistentVolumeClaim](#persistentvolumeclaim)
```
We could split “in tree” into core types like the downward API vs legacy vendor plugins that are in-tree because of their age, but would now not be accepted (you'd be asked to have a good long think about using CSI, right?)

sftim commented:
@geoffcline is there anything I can do to help move this onwards? Maybe a Zoom call or similar?

kbhawkey commented:
👋 @geoffcline .
Just checking if there is more work to do on this PR?

chrisnegus commented:
> 👋 @geoffcline . Just checking if there is more work to do on this PR?

I was speaking to @geoffcline about this and I have some ideas about how to proceed. I'll have something soon if we can keep this open for a bit longer.
